src/storage/v2/inmemory/edge_type_property_index.cpp

```diff
@@ -0,0 +1,333 @@
+      ++next_it;
+
+      if (it->timestamp >= oldest_active_start_timestamp) {
+        it = next_it;
+        continue;
+      }
+
+      const bool vertices_deleted = it->from_vertex->deleted || it->to_vertex->deleted;
+      const bool has_next = next_it != edges_acc.end();
+
+      // When we update specific entries in the index, we don't delete the previous entry.
+      // The way they are removed from the index is through this check. The entries should
+      // be right next to each other(in terms of iterator semantics) and the older one
+      // should be removed here.
+      bool redundant_duplicate = has_next && it->value == next_it->value && it->from_vertex == next_it->from_vertex &&
+                                 it->to_vertex == next_it->to_vertex && it->edge == next_it->edge;
+      if (redundant_duplicate || vertices_deleted ||
+          !AnyVersionHasLabelProperty(*it->edge, specific_index.first.second, it->value,
+                                      oldest_active_start_timestamp) ||
+          !std::ranges::all_of(it->from_vertex->out_edges, [&](const auto &edge) {
+            auto *to_vertex = std::get<InMemoryEdgeTypePropertyIndex::kVertexPos>(edge);
+            return to_vertex != it->to_vertex;
+          })) {
+        edges_acc.remove(*it);
+      }
+
+      it = next_it;
+    }
+  }
+}
+
+void InMemoryEdgeTypePropertyIndex::UpdateOnSetProperty(Vertex *from_vertex, Vertex *to_vertex, Edge *edge,
+                                                        EdgeTypeId edge_type, PropertyId property, PropertyValue value,
+                                                        uint64_t timestamp) {
+  if (value.IsNull()) {
+    return;
+  }
+
+  auto it = index_.find({edge_type, property});
+  if (it == index_.end()) return;
+
+  auto acc = it->second.access();
+  acc.insert({value, from_vertex, to_vertex, edge, timestamp});
+}
+
+uint64_t InMemoryEdgeTypePropertyIndex::ApproximateEdgeCount(EdgeTypeId edge_type, PropertyId property) const {
+  if (auto it = index_.find({edge_type, property}); it != index_.end()) {
+    return it->second.size();
+  }
+  return 0U;
+}
+
+void InMemoryEdgeTypePropertyIndex::UpdateOnEdgeModification(Vertex *old_from, Vertex *old_to, Vertex *new_from,
+                                                             Vertex *new_to, EdgeRef edge_ref, EdgeTypeId edge_type,
+                                                             PropertyId property, const Transaction &tx) {
+  auto it = index_.find({edge_type, property});
+  if (it == index_.end()) {
+    return;
+  }
+  auto acc = it->second.access();
+
+  auto entry_to_update = std::ranges::find_if(acc, [&](const auto &entry) {
+    return entry.from_vertex == old_from && entry.to_vertex == old_to && entry.edge == edge_ref.ptr;
+  });
+  const auto value = entry_to_update->value;
+  acc.remove(Entry{value, entry_to_update->from_vertex, entry_to_update->to_vertex, entry_to_update->edge,
+                   entry_to_update->timestamp});
```

Comment:
I think we shouldn't remove here this entry, what if some other transaction is using that index, and that edge is still visible for it?

Reply:
Agree. Since we are under guards, maybe we can update the value via find + updating fields using iterator?
I checked how do we do with label properties indices, it looks like we postpone cleaning operation and collect all garbage later.

Reply:
I wouldn't update fields in the entry, as maybe some other transaction will look for that entry. Correct thing would be to insert new entry just

Reply:
But in this case we need to collect garbage, right? Hm, will think how to improve there and check how do we do this for label-property indices.

Reply:
Not yet, what if this transaction that inserted index aborts? Then old entry is correct one, right

Reply:
resolved on call